### PR TITLE
refactor interactive learning to use new online infrastructure

### PR DESCRIPTION
**Pull request by ronuchit**
wip for interactive online stuff
* now that we're doing online learning without any hacks, we can completely get rid of `interactive_known_predicates` and should use the existing `excluded_predicates` infrastructure instead. i've done that in this PR
* removed hacks surrounding predicate stripping from interactive_learning_approach, now there's no way it can cheat. the initial predicates that are passed into `__init__` are the known predicates. the ones to be learned are extracted from the offline dataset at the start of `learn_from_offline_dataset` (note their classifiers were stripped during dataset construction, so there's no way to cheat)
* there's one functional difference that i wrote a "TODO" comment for, we should discuss and make a decision

@tomsilver @amburger66 read this before looking at the diffs!

### CODE REVIEW

File: src/main.py
Comment (ronuchit) [inline src/main.py:78]: @tomsilver This refactor moves this code to utils.py, since dataset `__init__.py` needs it too. Should we refactor this change into a separate PR? (Added benefit: test_main.py will get a nice speedup bc we don't need to run the whole pipeline just to test predicate exclusion lol)

Comment (tomsilver) [reply src/main.py:78]: sure sounds good to me

Comment (tomsilver) [reply src/main.py:78]: #380

File: src/approaches/interactive_learning_approach.py
Comment (tomsilver) [inline src/approaches/interactive_learning_approach.py:83]: good point. I think we do want to allow the agent to ask about multiple atoms per state. we could do what you suggest here, or we could change the Query/Response for this project to be about sets of ground atoms instead of one. I don't have a preference.

File: src/approaches/interactive_learning_approach.py
Comment (tomsilver) [inline src/approaches/interactive_learning_approach.py:74]: this is a significant functional change, but since this score function is just a placeholder anyway, I think this change is okay. it will certainly lead to more interesting and testable behavior, where queries continue after the first state. @amburger66 let us know what you think though.

**Comment by tomsilver** [comment]
TODO: we realized that ground-truth predicate classifiers are leaking into the interactive learning approach for goal predicates, through train tasks. we'll need to fix this either here or in a separate PR.

File: src/approaches/interactive_learning_approach.py
Comment (amburger66) [reply src/approaches/interactive_learning_approach.py:74]: sounds good!

Comment (tomsilver) [reply src/approaches/interactive_learning_approach.py:74]: cool. we'll still do this in a separate PR if we do it, to keep this one a pure refactor.

File: src/approaches/interactive_learning_approach.py
Comment (tomsilver) [reply src/approaches/interactive_learning_approach.py:83]: I decided to go with the latter: #381

**Comment by tomsilver** [comment]
There are a number of functional changes I'd like to make after we merge this PR, I'm putting them here so I don't forget:
* Change the annotations from positive only to positive/negative, and put the negative results from active learning into negative
* Implement an option to use a different kind of trivial score function for selecting atoms to ask about, one that just asks about everything. (Maybe change to >= instead of > in this step)
* Allow falling back to random policy if no goal can be found
* Figure out if we want the agent to be able to ask a different kind of query like "tell me everything in this state" or "tell me something in this state", which we would need to pure active learning, because otherwise it doesn't know what the universe of predicates is
* Test that active learning alone works

File: tests/approaches/test_interactive_approach.py
Comment (tomsilver) [inline tests/approaches/test_interactive_approach.py:64]: I'm removing this test because the expected behavior is different now; the agent will not attempt to learn about a predicate that it has no data about, because it won't even know that this predicate exists.

**Comment by tomsilver** [comment]
@amburger66 @ronuchit I think this is ready for review now.

After we merge, we should launch supercloud experiments to identify if we've introduced any major regressions. (@amburger66 , could you do this? thanks!)

File: tests/approaches/test_interactive_approach.py
Comment (ronuchit) [inline tests/approaches/test_interactive_approach.py:87]: oh lol so we knew about this issue before. we should def fix it!

File: src/approaches/interactive_learning_approach.py
Comment (ronuchit) [inline src/approaches/interactive_learning_approach.py:82]: should we change "if" to "when"? oh nvm, because the option model could be learned/wrong in the final version of the system (just keeping this comment for clarity's sake, no action needed)

File: src/approaches/interactive_learning_approach.py
Comment (ronuchit) [inline src/approaches/interactive_learning_approach.py:77]: note that we don't really need this... can always make the Query with an empty set. not sure if you'd prefer that idk

File: src/approaches/interactive_learning_approach.py
Comment (ronuchit) [inline src/approaches/interactive_learning_approach.py:98]: remove this line, we no longer need it (it was initially there to appease mypy when we were using `response.query`)

**Review by ronuchit — COMMENTED** [review]
woohoo! i can't approve as the author, minor comments

**Comment by amburger66** [comment]
> After we merge, we should launch supercloud experiments to identify if we've introduced any major regressions. (@amburger66 , could you do this? thanks!)

yup

**Comment by amburger66** [comment]
lgtm

**Review by tomsilver — APPROVED** [review]

**Pull request merged**
